### PR TITLE
Improve 'insertForJoin' function performance by group prefetching

### PR DESCRIPTION
**Pull request by zhli1142015**
Apply the prefetching optimization for join probe to function 'insertForJoin' to improve it's performance.
Fixes: #9732

### CODE REVIEW

**Comment by zhli1142015** [comment]
Verified with join fuzzer for 30 mins.
```
W0508 12:12:09.084839 3634699 JoinFuzzer.cpp:1232] ==============================> Done with iteration 9585
[==========] Running 0 tests from 0 test suites.
[==========] 0 tests from 0 test suites ran. (0 ms total)
[  PASSED  ] 0 tests.
```

File: velox/exec/HashTable.cpp [hunk 1]
```diff
@@ -1206,6 +1208,45 @@ FOLLY_ALWAYS_INLINE void HashTable<ignoreNullKeys>::buildFullProbe(
...
+template <bool ignoreNullKeys>
+template <bool isNormailizedKeyMode>
+FOLLY_ALWAYS_INLINE void HashTable<ignoreNullKeys>::insertForJoinInternal(
+    RowContainer* rows,
+    char** groups,
+    uint64_t* hashes,
+    int32_t numGroups,
+    TableInsertPartitionInfo* partitionInfo) {
+  auto i = 0;
+  constexpr int32_t groupSize = 64;
```
[the remark this answers is not in the document]
Comment (zhli1142015) [reply velox/exec/HashTable.cpp:1220]: I tried 32 / 64 / 128 on TPCDS, and also I tuned this for join probe, please see https://github.com/facebookincubator/velox/pull/6695#issuecomment-1736657327.
Let me test this parameter with my benchmark also.

File: velox/exec/HashTable.cpp [hunk 2]
```diff
@@ -1167,11 +1170,10 @@ FOLLY_ALWAYS_INLINE void HashTable<ignoreNullKeys>::buildFullProbe(
     storeRowPointer(index, hash, inserted);
     return nullptr;
   };
-
-  if (hashMode_ == HashMode::kNormalizedKey) {
+  if constexpr (isNormailizedKeyMode) {
```
[the remark this answers is not in the document]
Comment (zhli1142015) [reply velox/exec/HashTable.cpp:1173]: Let me do the comparsion prefetch VS. remove branching with my benchmark. Thanks.

File: velox/exec/HashTable.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (Yuhta) [reply velox/exec/HashTable.cpp:1220]: Let's put this constant at global scope if the same value is optimal for both build and probe

[the remark this answers is not in the document]
Comment (zhli1142015) [reply velox/exec/HashTable.cpp]: Updated, thanks.

[the remark this answers is not in the document]
Comment (zhli1142015) [reply velox/exec/HashTable.cpp]: 32
```
============================================================================
[...]HashJoinPrepareJoinTableBenchmark.cpp     relative  time/iter   iters/s
============================================================================
Size:100000,parallel:false,withDup:false,NORMAL             1.23ms    812.07
Size:100000,parallel:true,withDup:false,NORMALI             3.72ms    268.64
Size:100000,parallel:false,withDup:true,NORMALI             1.25ms    800.75
Size:100000,parallel:true,withDup:true,NORMALIZ             1.83ms    547.17
Size:2097149,parallel:false,withDup:false,NORMA            40.76ms     24.54
Size:2097149,parallel:true,withDup:false,NORMAL            18.83ms     53.11
Size:2097149,parallel:false,withDup:true,NORMAL            45.22ms     22.11
Size:2097149,parallel:true,withDup:true,NORMALI            20.70ms     48.30
Size:16777216,parallel:false,withDup:false,NORM           550.27ms      1.82
Size:16777216,parallel:true,withDup:false,NORMA           191.07ms      5.23
Size:16777216,parallel:false,withDup:true,NORMA           593.11ms      1.69
Size:16777216,parallel:true,withDup:true,NORMAL           159.83ms      6.26
Size:2097149,parallel:false,withDup:false,HASH             48.15ms     20.77
Size:2097149,parallel:true,withDup:false,HASH              23.98ms     41.71
Size:2097149,parallel:false,withDup:true,HASH              46.72ms     21.40
Size:2097149,parallel:true,withDup:true,HASH               19.00ms     52.64
Size:16777216,parallel:false,withDup:false,HASH           615.37ms      1.63
Size:16777216,parallel:true,withDup:false,HASH            201.70ms      4.96
Size:16777216,parallel:false,withDup:true,HASH            608.83ms      1.64
Size:16777216,parallel:true,withDup:true,HASH             198.92ms      5.03
```

128
```
============================================================================
[...]HashJoinPrepareJoinTableBenchmark.cpp     relative  time/iter   iters/s
============================================================================
Size:100000,parallel:false,withDup:false,NORMAL             1.27ms    786.02
Size:100000,parallel:true,withDup:false,NORMALI             4.35ms    229.85
Size:100000,parallel:false,withDup:true,NORMALI             1.18ms    848.35
Size:100000,parallel:true,withDup:true,NORMALIZ             1.88ms    532.44
Size:2097149,parallel:false,withDup:false,NORMA            42.53ms     23.51
Size:2097149,parallel:true,withDup:false,NORMAL            18.85ms     53.06
Size:2097149,parallel:false,withDup:true,NORMAL            43.08ms     23.21
Size:2097149,parallel:true,withDup:true,NORMALI            16.52ms     60.54
Size:16777216,parallel:false,withDup:false,NORM           607.19ms      1.65
Size:16777216,parallel:true,withDup:false,NORMA           165.20ms      6.05
Size:16777216,parallel:false,withDup:true,NORMA           602.84ms      1.66
Size:16777216,parallel:true,withDup:true,NORMAL           169.28ms      5.91
Size:2097149,parallel:false,withDup:false,HASH             54.27ms     18.42
Size:2097149,parallel:true,withDup:false,HASH              20.39ms     49.05
Size:2097149,parallel:false,withDup:true,HASH              49.66ms     20.14
Size:2097149,parallel:true,withDup:true,HASH               21.21ms     47.15
Size:16777216,parallel:false,withDup:false,HASH           649.31ms      1.54
Size:16777216,parallel:true,withDup:false,HASH            208.54ms      4.80
Size:16777216,parallel:false,withDup:true,HASH            653.28ms      1.53
Size:16777216,parallel:true,withDup:true,HASH             200.35ms      4.99

```

File: velox/exec/HashTable.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (zhli1142015) [reply velox/exec/HashTable.cpp:1175]: prefetch only

```
============================================================================
[...]HashJoinPrepareJoinTableBenchmark.cpp     relative  time/iter   iters/s
============================================================================
Size:100000,parallel:false,withDup:false,NORMAL             1.27ms    784.89
Size:100000,parallel:true,withDup:false,NORMALI             3.44ms    290.86
Size:100000,parallel:false,withDup:true,NORMALI             1.20ms    829.95
Size:100000,parallel:true,withDup:true,NORMALIZ             1.83ms    545.50
Size:2097149,parallel:false,withDup:false,NORMA            43.14ms     23.18
Size:2097149,parallel:true,withDup:false,NORMAL            18.82ms     53.14
Size:2097149,parallel:false,withDup:true,NORMAL            46.10ms     21.69
Size:2097149,parallel:true,withDup:true,NORMALI            19.16ms     52.19
Size:16777216,parallel:false,withDup:false,NORM           581.32ms      1.72
Size:16777216,parallel:true,withDup:false,NORMA           183.41ms      5.45
Size:16777216,parallel:false,withDup:true,NORMA           582.94ms      1.72
Size:16777216,parallel:true,withDup:true,NORMAL           171.51ms      5.83
Size:2097149,parallel:false,withDup:false,HASH             49.79ms     20.08
Size:2097149,parallel:true,withDup:false,HASH              25.15ms     39.76
Size:2097149,parallel:false,withDup:true,HASH              52.77ms     18.95
Size:2097149,parallel:true,withDup:true,HASH               25.80ms     38.75
Size:16777216,parallel:false,withDup:false,HASH           621.92ms      1.61
Size:16777216,parallel:true,withDup:false,HASH            199.66ms      5.01
Size:16777216,parallel:false,withDup:true,HASH            623.70ms      1.60
Size:16777216,parallel:true,withDup:true,HASH             206.67ms      4.84
```

remove branch only

```
============================================================================
[...]HashJoinPrepareJoinTableBenchmark.cpp     relative  time/iter   iters/s
============================================================================
Size:100000,parallel:false,withDup:false,NORMAL             1.14ms    876.16
Size:100000,parallel:true,withDup:false,NORMALI             3.74ms    267.54
Size:100000,parallel:false,withDup:true,NORMALI             1.06ms    946.51
Size:100000,parallel:true,withDup:true,NORMALIZ             1.86ms    538.48
Size:2097149,parallel:false,withDup:false,NORMA            75.46ms     13.25
Size:2097149,parallel:true,withDup:false,NORMAL            21.74ms     46.00
Size:2097149,parallel:false,withDup:true,NORMAL            71.18ms     14.05
Size:2097149,parallel:true,withDup:true,NORMALI            23.23ms     43.05
Size:16777216,parallel:false,withDup:false,NORM           734.93ms      1.36
Size:16777216,parallel:true,withDup:false,NORMA           211.57ms      4.73
Size:16777216,parallel:false,withDup:true,NORMA           721.20ms      1.39
Size:16777216,parallel:true,withDup:true,NORMAL           211.38ms      4.73
Size:2097149,parallel:false,withDup:false,HASH             74.45ms     13.43
Size:2097149,parallel:true,withDup:false,HASH              28.01ms     35.70
Size:2097149,parallel:false,withDup:true,HASH              69.74ms     14.34
Size:2097149,parallel:true,withDup:true,HASH               28.62ms     34.94
Size:16777216,parallel:false,withDup:false,HASH           796.38ms      1.26
Size:16777216,parallel:true,withDup:false,HASH            239.07ms      4.18
Size:16777216,parallel:false,withDup:true,HASH            782.28ms      1.28
Size:16777216,parallel:true,withDup:true,HASH             234.01ms      4.27
```

**Review by Yuhta — APPROVED** [review]

**Pull request closed without merging**